ansible/roles/test/files/ptftests/vlan_test.py

```diff
@@ -0,0 +1,268 @@
+            for permit_vlanid in map(int, vlan_port["permit_vlanid"].keys()):
+                self.send_icmp_packet(vlan_port, permit_vlanid)
+                self.verify_icmp_packets(vlan_port, permit_vlanid)
+
+        # Test case #3
+        # Send packets with invalid VLAN ID
+        # Verify no port can receive these pacekts
+        self.log("Test case #3 starting ...")
+        invalid_tagged_pkt = self.build_icmp_packet(4095)
+        masked_invalid_tagged_pkt = Mask(invalid_tagged_pkt)
+        masked_invalid_tagged_pkt.set_do_not_care_scapy(scapy.Dot1Q, "vlan")
+
+        for vlan_port in vlan_ports_list:
+            src_port = vlan_port["port_index"]
+            dst_ports = [port["port_index"] for port in vlan_ports_list 
+                                 if port != vlan_port ]
+            self.log("Send invalid tagged packet " + " from " + str(src_port) + "...")
+            send(self, src_port, invalid_tagged_pkt)
+            self.log("Check on " + str(dst_ports) + "...")
+            verify_no_packet_any(self, masked_invalid_tagged_pkt, dst_ports)
+
+        # Test case #4
+        # Send packets over VLAN interfaces. 
+        # Verify packets can be receive on the egress port.
+        self.log("Test case #4 starting ...")
+
+        target_list = []
+        for vlan_port in vlan_ports_list:
+            for vlan_id in vlan_port["permit_vlanid"].keys():
+                item = {"vlan_id": int(vlan_id), "port_index": vlan_port["port_index"], 
+                        "peer_ip": vlan_port["permit_vlanid"][vlan_id]["peer_ip"], 
+                        "remote_ip": vlan_port["permit_vlanid"][vlan_id]["remote_ip"],
+                        "pvid": vlan_port["pvid"]}
+                target_list.append(item)
+
+        for vlan_port in vlan_ports_list:
+            src_port = vlan_port["port_index"]
+            src_mac = self.dataplane.get_mac(0, src_port)
+            dst_mac = self.router_mac
+            for vlan_id in map(int, vlan_port["permit_vlanid"].keys()):
+                # Test for for directly-connected routing
+                src_ip = vlan_port["permit_vlanid"][str(vlan_id)]["peer_ip"]
+                for target in target_list:
+                    if vlan_id == target["vlan_id"]:
+                        # Skip same VLAN forwarding
+                        continue
+                    self.send_icmp_packet(vlan_port, vlan_id if vlan_id != vlan_port["pvid"] else 0,
+                                          src_mac, dst_mac, src_ip, target["peer_ip"])
+                    self.verify_icmp_packets_from_specified_port(target["port_index"], 
+                                                                 target["vlan_id"] if target["vlan_id"] != target["pvid"] else 0,
+                                                                 dst_mac, self.dataplane.get_mac(0, target["port_index"]),
+                                                                 src_ip, target["peer_ip"], 63)
```

Comment:
port 16 is part of lag, the packet can go to either port 16 or port 17. Here you only check port 16 which is not enough.
```
01:36:21.910  root      : INFO    : Send tagged(200) packet from 4...
01:36:21.910  root      : INFO    : 7c:fe:90:5e:6b:04 -> 90:b1:1c:f4:a8:53, 192.168.200.4 -> 192.168.100.16
01:36:21.912  root      : INFO    : Verify packet from port 16
```

Reply:
The test only supports t0 testbed type now. The t0 testbed type should only suppprt one member in LAG? If t1-lag is supported in the future, here need to be modified indeed.

Reply:
```
05:14:27.515  root      : INFO    : Test case #4 starting ...
05:14:27.524  root      : INFO    : Send tagged(200) packet from 4...
05:14:27.525  root      : INFO    : 7c:fe:90:5e:6b:04 -> 90:b1:1c:f4:a8:53, 192.168.200.4 -> 192.168.100.4
05:14:27.528  root      : INFO    : Verify packet from port 4
05:14:27.543  root      : INFO    : Send tagged(200) packet from 4...
05:14:27.543  root      : INFO    : 7c:fe:90:5e:6b:04 -> 90:b1:1c:f4:a8:53, 192.168.200.4 -> 192.168.100.16
05:14:27.545  root      : INFO    : Verify packet from port 16
05:14:27.558  root      : INFO    : Send tagged(200) packet from 4...
05:14:27.559  root      : INFO    : 7c:fe:90:5e:6b:04 -> 90:b1:1c:f4:a8:53, 192.168.200.4 -> 192.168.100.0
05:14:27.562  root      : INFO    : Verify packet from port 0
````

The destination IP 192.168.100.0 is strange